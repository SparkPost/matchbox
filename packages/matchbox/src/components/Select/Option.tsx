import React from 'react';

export type SelectOptionProps =
  | string
  | number
  | ({
      value: string | number;
      label: string;
    } & React.ComponentPropsWithRef<'option'>);

const Option = ({ option }: { option: SelectOptionProps }): JSX.Element => {
  if (typeof option === 'object') {
    const { value, label = value, ...rest } = option;
    return (
      <option value={value} {...rest}>
        {label}
      </option>
    );
  } else if (typeof option === 'string' || typeof option === 'number') {
    return <option value={option}>{option}</option>;
  }
};

export default Option;
