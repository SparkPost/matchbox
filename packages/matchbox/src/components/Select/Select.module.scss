@import '../../styles/config';

.Select {
  position: relative;
  border: none;
  padding: 0;
  margin: 0 0 spacing() 0;

  &:last-child {
    margin: 0;
  }
}

.error .Input {
  border: 1px solid color(red);
  background: rgba(color(red, dark), 0.06);
}

.InlineError {
  padding-left: spacing(small);
}

.Input {
  appearance: none;
  position: relative;
  display: block;
  box-sizing: border-box;
  width: 100%;

  padding: rem(5) rem(27) rem(5) rem(12);

  outline: none;
  border: 1px solid color(gray, 7);
  border-radius: border-radius();
  background: linear-gradient(180deg, color(gray, 10), color(gray, 9));

  color: color(gray, 1);
  font-weight: 400;
  font-size: font-size(400);
  line-height: rem(24);

  box-shadow: shadow();

  transition: border 0.15s;

  &:hover {
    cursor: pointer;
  }

  &:focus {
    border: 1px solid color(blue);
    box-shadow: 0 0 0 1px color(blue);
  }

  &:disabled {
    background: color(gray, 9);
    color: color(gray, 5);
    cursor: not-allowed;
  }
}

// Firefox doesn't respect line-height of <select>
// @see https://bugzilla.mozilla.org/show_bug.cgi?id=1241167
@-moz-document url-prefix() {
  .Input {
    height: rem(36);
  }
}

.Dropdown {
  position: absolute;
  z-index: 1;
  top: rem(30);
  right: rem(9);
  pointer-events: none;
  fill: color(gray, 4);
}

.labelHidden {
  top: rem(9);
}

.HelpText {
  padding-top: rem(6);
  font-size: font-size(300);
  line-height: line-height(300);
  color: color(gray, 4);
}
