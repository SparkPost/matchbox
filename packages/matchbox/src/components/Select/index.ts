export { default as Select } from './Select';
export * from './Select';
export * from './Options';
export * from './Option';
