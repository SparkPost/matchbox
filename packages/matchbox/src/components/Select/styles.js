import { tokens } from '@sparkpost/design-tokens';

export const select = props => `
  appearance: none;
  &:hover {
    cursor: ${props.disabled ? 'not-allowed;' : 'pointer;'}
  }
`;

export const chevron = props => `
  position: absolute;
  z-index: 1;
  right: ${tokens.spacing_300};
  height: 100%;
  fill: ${tokens.color_blue_700};
  user-select: none;
  pointer-events: none;
  ${props.disabled ? 'display: none' : ''}
`;
