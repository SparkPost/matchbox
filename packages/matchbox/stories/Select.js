import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Select } from '../src';

const options = ['Foo', 2, 'Bar'];
const options2 = [
  {
    value: 1,
    label: 'One',
    pass: 'through'
  },
  {
    value: '2',
    label: 'Two',
  }
];

export default storiesOf('Select', module)
  .addDecorator((getStory) => (
    <StoryContainer bg='white'>{ getStory() }</StoryContainer>
  ))
  .add('Default', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      placeholder='Leslie Knope'
      defaultValue=''
      options={options2}
    />
  )))

  .add('with strings or numbers', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      placeholder='Leslie Knope'
      options={options}
    />
  )))

  .add('with an error', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      options={options2}
      error='Your forgot to select'
    />
  )))

  .add('disabled', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      options={options2}
      disabled
    />
  )))

  .add('required', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      options={options2}
      required
    />
  )))

  .add('with help text', withInfo()(() => (
    <Select
      id='id'
      label='Select an option'
      options={options2}
      helpText="Remember to select something"
    />
  )))

  .add('with custom value for placeholder', withInfo()(() => (
    <Select
      placeholderValue="NONE"
      placeholder='Select one'
      defaultValue="NONE"
      options={options2}
    />
  )));
