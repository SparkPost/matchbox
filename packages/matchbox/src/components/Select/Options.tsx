import React from 'react';
import Option, { SelectOptionProps } from './Option';

export type SelectOptionsProps = {
  options: SelectOptionProps[];
  placeholderValue?: string;
  placeholder: string;
};

function Options({ options, placeholder, placeholderValue }: SelectOptionsProps) {
  let combined = options;

  if (placeholder) {
    combined = [{ label: placeholder, value: placeholderValue, disabled: true }, ...combined];
  }

  if (!combined && !combined.length) {
    return null;
  }

  return (
    <>
      {combined.map((option, key) => (
        <Option option={option} key={key} />
      ))}
    </>
  );
}

export default Options;
