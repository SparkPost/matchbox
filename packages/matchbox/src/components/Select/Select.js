import React from 'react';
import PropTypes from 'prop-types';
import { tokens } from '@sparkpost/design-tokens';
import { Label } from '../Label';
import { Error } from '../Error';
import { KeyboardArrowDown } from '@sparkpost/matchbox-icons';
import styled from 'styled-components';
import { select, chevron } from './styles';
import { Box } from '../Box';
import { OptionalLabel } from '../OptionalLabel';
import { HelpText } from '../HelpText';
import { compose, margin, maxWidth } from 'styled-system';
import { createPropTypes } from '@styled-system/prop-types';
import { omit } from '@styled-system/props';
import { pick } from '../../helpers/props';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { focusOutline } from '../../styles/helpers';

const Option = ({ option }) => {
  if (typeof option === 'object') {
    const { value, label = value, ...rest } = option;
    return (
      <option value={value} {...rest}>
        {label}
      </option>
    );
  } else if (typeof option === 'string' || typeof option === 'number') {
    return <option value={option}>{option}</option>;
  }
};

function Options({ options, placeholder, placeholderValue }) {
  let combined = options;

  if (placeholder) {
    combined = [{ label: placeholder, value: placeholderValue, disabled: true }, ...combined];
  }

  if (!combined && !combined.length) {
    return null;
  }

  return (
    <>
      {combined.map((option, key) => (
        <Option option={option} key={key} />
      ))}
    </>
  );
}

const StyledInputWrapper = styled(Box)`
  ${focusOutline({ modifier: '&:focus-within', offset: '2px' })}
`;

const StyledInput = styled(Box)`
  outline: none;
`;

const SelectBox = React.forwardRef(function SelectBox(props, userRef) {
  const { hasError, disabled, ...rest } = props;
  return (
    <StyledInputWrapper>
      <StyledInput
        aria-invalid={!!hasError}
        as="select"
        disabled={disabled}
        width="100%"
        border={hasError ? `1px solid ${tokens.color_red_700}` : '400'}
        borderRadius="100"
        bg={disabled ? 'gray.200' : 'white'}
        pl="400"
        pr="600"
        lineHeight="2.5rem"
        height="2.5rem"
        color="gray.900"
        {...rest}
        ref={userRef}
      />
    </StyledInputWrapper>
  );
});

const system = compose(margin, maxWidth);

const StyledSelect = styled(SelectBox)`
  ${select}
`;

const StyledChevron = styled(KeyboardArrowDown)`
  ${chevron}
`;

const StyledWrapper = styled('div')`
  ${system}
`;

const Select = React.forwardRef(function Select(props, userRef) {
  const {
    id,
    options,
    label,
    helpText,
    placeholder,
    placeholderValue,
    disabled,
    required,
    error,
    errorInLabel,
    optional,
    // labelHidden, TODO add this back in later after hibana cutover
    ...rest
  } = props;
  const systemProps = pick(rest, system.propNames);
  const componentProps = omit(rest);
  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  const requiredIndicator = required ? (
    <Box as="span" pr="200" aria-hidden="true">
      *
    </Box>
  ) : null;

  const labelMarkup = (
    <Label
      id={id}
      label={label}
      // labelHidden={labelHidden} TODO Add this back in after hibana cutover
    >
      {requiredIndicator}
      {error && errorInLabel && (
        <Box as={Error} id={errorId} wrapper="span" error={error} fontWeight="400" />
      )}
      {optional && <OptionalLabel float />}
    </Label>
  );

  const helpMarkup = helpText ? <HelpText id={helpTextId}>{helpText}</HelpText> : null;

  return (
    <StyledWrapper {...systemProps}>
      {labelMarkup}
      <Box position="relative">
        <StyledSelect
          id={id}
          disabled={disabled}
          required={required}
          hasError={!!error}
          {...componentProps}
          {...describedBy}
          ref={userRef}
        >
          <Options
            options={options}
            placeholder={placeholder}
            placeholderValue={placeholderValue}
          />
        </StyledSelect>
        <StyledChevron size={24} disabled={disabled} />
      </Box>
      {helpMarkup}
      {error && !errorInLabel && <Error id={errorId} error={error} />}
    </StyledWrapper>
  );
});

Select.displayName = 'Select';
Select.propTypes = {
  id: PropTypes.string,
  /**
   * Select options -
   * Array of Objects with { value, label }, Strings, or Numbers
   */
  options: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.shape({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        label: PropTypes.string.isRequired,
      }),
      PropTypes.string.isRequired,
      PropTypes.number.isRequired,
    ]),
  ).isRequired,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  label: PropTypes.string,
  // labelHidden: PropTypes.bool, TODO Add this back in after hibana cutover
  helpText: PropTypes.node,
  error: PropTypes.string,
  errorInLabel: PropTypes.bool,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  optional: PropTypes.bool,
  ...createPropTypes(margin.propNames),
  ...createPropTypes(maxWidth.propNames),
};

export default Select;
