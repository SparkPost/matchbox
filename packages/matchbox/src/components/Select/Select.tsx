import React from 'react';
import { tokens } from '@sparkpost/design-tokens';
import { Label } from '../Label';
import { ErrorLabel } from '../Error';
import { KeyboardArrowDown } from '@sparkpost/matchbox-icons';
import styled from 'styled-components';
import { select, chevron } from './styles';
import { Box, BoxProps } from '../Box';
import { OptionalLabel } from '../OptionalLabel';
import { HelpText } from '../HelpText';
import { compose, margin, MarginProps, maxWidth, MaxWidthProps } from 'styled-system';
import { omit } from '@styled-system/props';
import { pick } from '../../helpers/props';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { focusOutline } from '../../styles/helpers';
import Options from './Options';
import { SelectOptionProps } from './Option';

const StyledInputWrapper = styled(Box)<BoxProps>`
  ${focusOutline({ modifier: '&:focus-within', offset: '2px' })}
`;

const StyledInput = styled(Box)<BoxProps>`
  outline: none;
`;

type SelectBoxProps = {
  hasError?: boolean;
  disabled?: boolean;
};

const SelectBox = React.forwardRef<HTMLSelectElement, SelectBoxProps>(function SelectBox(
  props,
  userRef,
) {
  const { hasError, disabled, ...rest } = props;
  return (
    <StyledInputWrapper>
      <StyledInput
        aria-invalid={!!hasError}
        as="select"
        disabled={disabled}
        width="100%"
        border={hasError ? `1px solid ${tokens.color_red_700}` : '400'}
        borderRadius="100"
        bg={disabled ? 'gray.200' : 'white'}
        pl="400"
        pr="600"
        lineHeight="2.5rem"
        height="2.5rem"
        color="gray.900"
        {...rest}
        ref={userRef}
      />
    </StyledInputWrapper>
  );
});

const system = compose(margin, maxWidth);

const StyledSelect = styled(SelectBox)`
  ${select}
`;

const StyledChevron = styled(KeyboardArrowDown)<{ size?: number; $disabled?: boolean }>`
  ${chevron}
`;

const StyledWrapper = styled('div')`
  ${system}
`;

export type SelectProps = {
  options: SelectOptionProps[];
  label?: string;
  labelHidden?: boolean;
  helpText?: React.ReactNode;
  error?: string;
  errorInLabel?: boolean;
  optional?: boolean;
  placeholderValue?: string;
} & MarginProps &
  MaxWidthProps &
  React.ComponentPropsWithRef<'select'>;

const Select = React.forwardRef<HTMLSelectElement, SelectProps>(function Select(props, userRef) {
  const {
    id,
    options,
    label,
    helpText,
    placeholder,
    placeholderValue,
    disabled,
    required,
    error,
    errorInLabel,
    optional,
    labelHidden,
    ...rest
  } = props;
  const systemProps = pick(rest, system.propNames);
  const componentProps = omit(rest);
  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  const labelMarkup = label && (
    <Label id={id} labelHidden={labelHidden}>
      <Box as="span" pr="200">
        {label}
      </Box>
      {error && errorInLabel && (
        <ErrorLabel id={errorId} wrapper="span">
          {error}
        </ErrorLabel>
      )}
      {optional && <OptionalLabel float />}
    </Label>
  );

  const helpMarkup = helpText ? <HelpText id={helpTextId}>{helpText}</HelpText> : null;

  return (
    <StyledWrapper {...systemProps}>
      {labelMarkup}
      <Box position="relative">
        <StyledSelect
          id={id}
          disabled={disabled}
          required={required}
          hasError={!!error}
          {...componentProps}
          {...describedBy}
          ref={userRef}
        >
          <Options
            options={options}
            placeholder={placeholder}
            placeholderValue={placeholderValue}
          />
        </StyledSelect>
        <StyledChevron size={24} $disabled={disabled} />
      </Box>
      {helpMarkup}
      {error && !errorInLabel && <ErrorLabel id={errorId}>{error}</ErrorLabel>}
    </StyledWrapper>
  );
});

Select.displayName = 'Select';

export default Select;
