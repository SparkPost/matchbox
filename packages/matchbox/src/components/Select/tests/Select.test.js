import React from 'react';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

import Select from '../Select';

describe('Select', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      options: ['option 1']
    };

    wrapper = shallow(<Select {...props} />);
  });

  const testCases = [
    { name: 'disabled', props: { disabled: true }},
    { name: 'required', props: { required: true, label: 'Select Label' }},
    { name: 'label', props: { label: 'Select Label' }},
    { name: 'withId', props: { id: 'with-an-id' }},
    { name: 'helpText', props: { helpText: 'Select help text' }},
    { name: 'error', props: { label: 'An error occurred' }},
    { name: 'errorInLabel', props: { error: 'Error occurred', errorInLabel: true }},
    { name: 'with options', props: { options: ['option 1', 'option 2']}}
  ];

  cases('renders Select', (opts) => {
    wrapper.setProps(opts.props);
    expect(wrapper).toMatchSnapshot();
  }, testCases);

  cases('invokes event', (opts) => {
    const fn = jest.fn();
    const newProps = {};
    newProps[opts.name] = fn;
    wrapper.setProps(newProps);
    wrapper.find('select').simulate(opts.event, opts.args); //due to enzyme no propagating event, it needs to trigger on exact element
    expect(fn).toHaveBeenCalledTimes(1);
  }, [
    { name: 'onChange', event: 'change' },
    { name: 'onBlur', event: 'blur' },
    { name: 'onFocus', event: 'focus' }
  ]);

  it('should render Option correctly', () => {
    wrapper.setProps({ options: ['option 1', 2, { label: 'three', value: 3, disabled: true } ]});
    const options = wrapper.find('Option');
    options.forEach((option) => expect(option.dive()).toMatchSnapshot());
  });
});
