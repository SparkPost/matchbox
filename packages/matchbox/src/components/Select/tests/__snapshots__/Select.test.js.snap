// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders Select disabled 1`] = `
<div
  className="Select"
>
  <select
    className="Input disabled"
    disabled={true}
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown labelHidden"
  />
</div>
`;

exports[`Select renders Select error 1`] = `
<div
  className="Select"
>
  <Label
    label="An error occurred"
  />
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown"
  />
</div>
`;

exports[`Select renders Select errorInLabel 1`] = `
<div
  className="Select error"
>
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown labelHidden"
  />
</div>
`;

exports[`Select renders Select helpText 1`] = `
<div
  className="Select"
>
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown labelHidden"
  />
  <div
    className="HelpText"
  >
    Select help text
  </div>
</div>
`;

exports[`Select renders Select label 1`] = `
<div
  className="Select"
>
  <Label
    label="Select Label"
  />
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown"
  />
</div>
`;

exports[`Select renders Select required 1`] = `
<div
  className="Select"
>
  <Label
    label="Select Label *"
  />
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown"
  />
</div>
`;

exports[`Select renders Select with options 1`] = `
<div
  className="Select"
>
  <select
    className="Input"
  >
    <Option
      key="0"
      option="option 1"
    />
    <Option
      key="1"
      option="option 2"
    />
  </select>
  <ArrowDropDown
    className="Dropdown labelHidden"
  />
</div>
`;

exports[`Select renders Select withId 1`] = `
<div
  className="Select"
>
  <select
    className="Input"
    id="with-an-id"
  >
    <Option
      key="0"
      option="option 1"
    />
  </select>
  <ArrowDropDown
    className="Dropdown labelHidden"
  />
</div>
`;

exports[`Select should render Option correctly 1`] = `
<option
  value="option 1"
>
  option 1
</option>
`;

exports[`Select should render Option correctly 2`] = `
<option
  value={2}
>
  2
</option>
`;

exports[`Select should render Option correctly 3`] = `
<option
  disabled={true}
  value={3}
>
  three
</option>
`;
